modulemd/include/private/modulemd-util.h

```diff
@@ -63,49 +129,195 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (modulemd_tracer, modulemd_trace_free);
 
 G_END_DECLS
 
-
+/**
+ * modulemd_hash_table_deep_str_copy:
+ * @orig: A #GHashTable to copy, containing string keys and values.
+ *
+ * Returns: A newly-allocated #GHashTable containing a deep copy of both the
+ * keys and values from @orig.
+ *
+ * Since: 2.0
+ */
 GHashTable *
 modulemd_hash_table_deep_str_copy (GHashTable *orig);
 
+/**
+ * modulemd_hash_table_deep_set_copy:
+ * @orig: A #GHashTable to copy, containing string keys.
+ *
+ * Returns: A newly-allocated #GHashTable containing a deep copy of the
+ * keys from @orig. The values from @orig are ignored, and the values in the
+ * copy are set the same as the corresponding keys so the returned #GHashTable
+ * can be used as a set.
+ *
+ * Since: 2.0
+ */
 GHashTable *
 modulemd_hash_table_deep_set_copy (GHashTable *orig);
 
+/**
+ * modulemd_hash_table_deep_str_set_copy:
+ * @orig: A #GHashTable to copy, containing string keys and #GHashTable values.
+ *
+ * Returns: A newly-allocated #GHashTable containing a deep copy of the
+ * keys from @orig. The corresponding #GHashTable value for each key is deep
+ * copied via modulemd_hash_table_deep_str_copy() for use as a set.
+ *
+ * TODO.
+ *
+ * Since: 2.0
+ */
 GHashTable *
 modulemd_hash_table_deep_str_set_copy (GHashTable *orig);
 
+/**
+ * modulemd_hash_table_deep_str_str_set_copy:
+ * @orig: A #GHashTable to copy, containing string keys and #GHashTable values
+ * that are nested two levels deep.
+ *
+ * Returns: A newly-allocated #GHashTable containing a deep copy of the
+ * keys from @orig. Each corresponding #GHashTable value is deep copied, with
+ * the second level #GHashTable copied for use as a set as described in
+ * modulemd_hash_table_deep_str_copy().
+ *
+ * Since: 2.0
+ */
 GHashTable *
 modulemd_hash_table_deep_str_str_set_copy (GHashTable *orig);
 
+/**
+ * modulemd_hash_table_sets_are_equal_wrapper:
```

Comment:
Note that this takes `(const void *)` to match the definition of a GEqualFunc. The function will fail and return FALSE if it is passed something other than a `GHashTable *`.

Reply:
_Will_ it fail gracefully if something other than `GHashTable *`s are passed? Or will it just explode or (even worse) work with bizarre effects. The function just casts the pointers and calls `modulemd_hash_table_sets_are_equal()` without checking their type. Is it even possible to check that the pointers are to `GHashTable`s?

Reply:
You’re right, we can’t actually handle this gracefully.